cmd/anonymizer/app/writer/writer.go

```diff
@@ -130,3 +139,9 @@ func (w *Writer) WriteSpan(msg *model.Span) error {
 
 	return nil
 }
+
+// Close closes the captured and anonymized files.
+func (w *Writer) Close() {
+	w.capturedFile.Close()
```

Comment:
you're missing things like `w.capturedFile.WriteString("\n]\n")`. See L132 - that code should be moved into Close.

Reply:
how did you test it? are you able to load the resulting fie in the UI?

Reply:
Hi @yurishkuro 

This is the output from a sample run: https://gist.github.com/Ashmita152/68904c2ffe2dc40bf17984ffc981cace
It doesn't load in the UI since the json isn't as expected I guess, no errors in js console:  https://gist.github.com/Ashmita152/e9b421819571098d3c9de965b6ae689c 
I am unable to figure out the reason about this inconsistency but my guess is that UI accepts a complete trace json while we are uploading one particular span json.

I have a question: why do we write span by span while writing to file, shouldn't we be first constructing a trace out of those spans and then write to file.

Reply:
ok, I see the confusion. The original purpose of the Writer was to use it while tailing a long stream of spans (in Kafka) to share them with researchers. It does not have the notion of a trace because spans are not ordered/grouped by trace ID in a stream. We had another internal binary that used these packages as a library.

For the purpose of this ticket, we know that we're dealing with a single trace. However, the Writer does not have the logic to write that trace in exactly the UI format, so you may need to add that. You can download a sample trace from the UI (via JSON button in the trace view) to see where the formatting is different.

Reply:
I see, that clears my doubt. Sure I will work on implementing that in writer. Do we need to keep the functionality backward compatible ie. allowing current tailing stream of spans too ?

Reply:
yes, please keep backward compatibility.

Reply:
I am confused about how to implement the logic to write the trace in the UI format. To me, it seems like we can't use WriteSpan function. I am thinking to do:
- Implement a WriteTrace function in writer.go
- Change QueryTrace in query.go which is now returning []model.Span. I am thinking to modify it to return model.Trace.
- In main.go instead of calling writer.writeSpan, call writer.writerTrace.
I would like to get your opinion before start implementing this way.

Reply:
yeah, the writer is not exactly compatible with saving a single trace. Let's merge this PR, since it achieved the basic anonymization of a trace, just doesn't output UI-compatible file, which can be added separately.

TBH, our internal program worked differently, in two stages - first we'd run it on a stream extracting lots of spans into an anonymized file. Then we had a separate command in the binary that would take a trace ID, scan the previously save file, pick the spans for the given trace ID, and output just that trace in the UI format. It would be weird to do it the same way here because the first run is already taking a trace ID as input (unlike our case where input was undifferentiated stream of spans from many traces). However, we could still implement it the same way, just invoke the second step automatically on the file already generated by your current code. 

Let me do it as a separate PR.

Reply:
#2629